mini-projects/01-device-driver/laptop-interface/src/controller.rs

```diff
@@ -0,0 +1,516 @@
+        let (output_sender, output_receiver) = std::sync::mpsc::channel();
+        ControllerManager {
+            controllers: Vec::new(),
+            input_receiver: output_receiver,
+            output_sender,
+            join_handles: HashMap::new(),
+        }
+    }
+
+    pub fn connect_controller(&mut self, serial_port: &str) -> Result<u32, ControllerError> {
+        let id = self.extract_id_from_serial_port(serial_port)?;
+
+        log::info!(
+            "Connecting to controller on serial port: {}, assigning id {}",
+            serial_port,
+            id
+        );
+
+        let (tx, rx): (Sender<String>, Receiver<String>) = std::sync::mpsc::channel();
+
+        let controller = Controller {
+            name: format!("Controller_{}", serial_port.replace("/", "_")),
+            id,
+            serial_port: serial_port.to_string(),
+            kind: ControllerManager::determine_controller_kind(serial_port),
+            state: ControllerState::default(),
+            sender: tx.clone(),
+        };
+
+        let controller_id = controller.id;
+
+        self.join_handles.insert(
+            controller_id,
+            Some(self.spawn_controller_thread(&controller, rx)?),
+        );
+        self.controllers.push(controller);
+        self.stop_controller(controller_id)?;
+        self.reset_controller(controller_id)?;
+        self.init_controller(controller_id)?;
+        self.set_controller_led(controller_id, LedState::Ready)?;
+
+        log::info!("Controller connected successfully");
+        Ok(controller_id)
+    }
+
+    fn spawn_controller_thread(
+        &self,
+        controller: &Controller,
+        rx: Receiver<String>,
+    ) -> Result<JoinHandle<Result<(), ControllerError>>, ControllerError> {
+        let thread_sender = self.output_sender.clone();
+        let controller_id = controller.id;
+        let serial_port = controller.serial_port.clone();
+
+        // Move the necessary data into the closure
+        Ok(thread::spawn(move || -> Result<(), ControllerError> {
+            log::trace!("Controller {}: Thread spawned.", controller_id);
+            // Initialize the serial port
+            let port = SerialPort::new(CString::new(serial_port.clone()).map_err(|e| {
+                log::error!(
+                    "Controller {}: Failed to convert serial port name to CString: {}",
+                    controller_id,
+                    e
+                );
+                ControllerError::SerialError(SerialError::Unknown)
+            })?)?;
+            log::trace!("Controller {}: SerialPort object created.", controller_id);
+
+            // Open the port
+            port.open(sp_mode::SP_MODE_READ_WRITE).map_err(|e| {
+                log::error!(
+                    "Controller {}: Failed to open serial port: {:?}",
+                    controller_id,
+                    e
+                );
+                ControllerError::SerialError(e)
+            })?;
```

Comment:
nit: I would think that establishing the serial interface would happen when the controller is connected to